airbyte-integrations/connectors/source-google-analytics-v4/source_google_analytics_v4/source.py

```diff
@@ -0,0 +1,493 @@
+            column_data_type = column_attributes.get("dataType")
+
+            if column_type == "METRIC":
+                metrics[column_name] = column_data_type
+            elif column_type == "DIMENSION":
+                dimensions[column_name] = column_data_type
+
+        return dimensions, metrics
+
+
+class GoogleAnalyticsV4Stream(HttpStream, ABC):
+    primary_key = None
+    http_method = "POST"
+    page_size = 100000
+    url_base = "https://analyticsreporting.googleapis.com/v4/"
+    report_field = "reports"
+    data_fields = ["data", "rows"]
+
+    map_type = dict(INTEGER="integer", FLOAT="number", PERCENT="number", TIME="number")
+
+    def __init__(self, config: Dict):
+        super().__init__(authenticator=config["authenticator"])
+        self.start_date = config["start_date"]
+        self.window_in_days = config["window_in_days"]
+        self.view_id = config["view_id"]
+        self.metrics = config["metrics"]
+        self.dimensions = config["dimensions"]
+        self._config = config
+        self.dimensions_ref, self.metrics_ref = GoogleAnalyticsV4TypesList().read_records(sync_mode=None)
+
+    @property
+    def state_checkpoint_interval(self) -> int:
+        return self.window_in_days
+
+    @staticmethod
+    def to_datetime_str(date: datetime) -> str:
+        """
+        Custom method.
+        Returns the formated datetime string.
+        :: Output example: '2021-07-15 07' FORMAT : "%Y-%m-%d"
+        """
+        return date.strftime("%Y-%m-%d")
+
+    def path(self, **kwargs) -> str:
+        return "reports:batchGet"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        next_page = response.json().get("nextPageToken")
+        if next_page:
+            return {"pageToken": next_page}
+
+    def request_body_json(
+        self, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> Optional[Mapping]:
+
+        metrics = [{"expression": metric} for metric in self.metrics]
+        dimensions = [{"name": dimension} for dimension in self.dimensions]
+
+        request_body = {
+            "reportRequests": [
+                {
+                    "viewId": self.view_id,
+                    "dateRanges": [stream_slice],
+                    "pageSize": self.page_size,
+                    "metrics": metrics,
+                    "dimensions": dimensions,
+                }
+            ]
+        }
+
+        if next_page_token:
+            request_body["reportRequests"][0].update(next_page_token)
+
+        return request_body
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        """
+        Override get_json_schema CDK method to retrieve the schema information for GoogleAnalyticsV4 Object dynamically.
+        """
+
+        schema = {"type": ["null", "object"], "additionalProperties": False, "properties": {"view_id": {"type": ["string"]}}}
```

Comment:
Make it 'multiline'


schema = {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": ["null", "object"], 
    "additionalProperties": False, 
    "properties": {
        "view_id": {
            "type": ["string"]
        },
    },
}

Reply:
Updated